SR2 Community Patch/scripts/server/components/Resources.as

```diff
@@ -6,17 +6,73 @@ import statuses;
 from resources import _tempResource;
 import bool getCheatsEverOn() from "cheats";
 
+// CE: Native Resource sorting logic to improve primaryResource identification
+// on planets with multiple tiered resources, main effect is mods which allow
+// stacking resources onto a single planet will work better, as vanilla doesn't
+// have any mechanisms for doing this with resources that actually matter
+// with the very minor exception of the core system of the Expanse, which
+// spawns planets with multiple tiered resources. For vanilla this change
+// ensures that Ancient planets will correctly be set to the level of the
+// highest spawned resource by making sure that becomes the primary one.
+const ResourceClass@ scalableClass;
+
 tidy class NativeResource : Resource {
 	TradePath@ path;
 
 	int opCmp(const NativeResource@ other) const {
-		return Resource::opCmp(other);
+		// Custom sorting logic to implement better primaryResource
+		// identification.
+		if (type is null && other.type is null)
+			return 0;
+		if (type is null)
+			return -1;
+		if (other.type is null)
+			return 1;
+		// compare resource types
+		int level = type.level;
+		int otherLevel = other.type.level;
+		if (scalableClass !is null) {
+			// treat scalable classes as far higher priority than
+			// tier classes, as they need to be the primary resource
+			// for functionality like Ancient Infusers
+			if (type.cls is scalableClass) {
+				level = 10;
+			}
+			if (other.type.cls is scalableClass) {
+				otherLevel = 10;
+			}
+		}
+		// order based on level before display weight or rarity
+		// to ensure adding a tier 2 resource to a tier 1 planet
+		// converts the primary resource to tier 2, which again is
+		// quite important for Ancient
+		if (level < otherLevel)
+			return -1;
+		if (level > otherLevel)
+			return 1;
+
+		// sort as Resource::opCmp would if no tie break yet
```

Comment:
I'm curious, why did you decide to copy the implementation of `Resource::opCmp(const Resource@)` when this was already how the game used to handle it?

Reply:
This is using the remaining part of the Resource compare method after the modified first part fails to tie break. If I called the vanilla compare method it would likely tie break on the fields I already compared with the refactored logic

Reply:
Oh, I see.